code/modules/clothing/glasses/glasses.dm

```diff
@@ -52,20 +52,21 @@
 	//In case someone in the future adds a non-toggle action to a child type
 	if(istype(action, /datum/action/item_action/toggle))
 		var/datum/action/item_action/toggle/toggle = action
-		toggle.toggled = !activate(user)
+		//Toggling item actions are dumb; it has to be done this way for the icon to properly update
+		toggle.set_toggle(activate(user))
 		return
 
-	activate(user)
+	return activate(user)
 
 ///Toggle the functions of the glasses
-/obj/item/clothing/glasses/proc/activate(mob/user, silent = FALSE)
-	if(!silent)
+/obj/item/clothing/glasses/proc/activate(mob/user)
+	if(activation_sound && deactivation_sound)
 		playsound(get_turf(src), active ? deactivation_sound : activation_sound, 15)
 
 	active = !active
+	update_icon_state()	//Found out the hard way this has to be before update_inv_glasses()
```

Comment:
don't call this directly

Reply:
What should I call then? Just `update_icon`?

Reply:
yes